spec/consensus/gossip/gossip.md

```diff
@@ -0,0 +1,333 @@
+
+Roughly speaking, if Gossip Communication is guaranteed, then all messages sent by correct processes will be eventually delivered to all correct processes and allow all correct processes to realize that the same conditions are.
+
+However, since processes are subject to failures, it is not guaranteed that proposals are ever sent and correct processes cannot wait indefinitely.
+Hence, processes execute in rounds in which they wait for conditions to be met for sometime and, if they timeout, send negative messages that will lead to new rounds.
+Again, Gossip Communication guarantees that eventually the conditions for deciding are met, even if only after GST is reached.
+
+Implementing Gossip Communication, however, is hard because even messages sent before the GST need to be buffered to be reliably delivered between correct processes, but since GST may take indefinitely long to arrive, in practice implementing this property would require unbounded memory.
+
+Fortunately, while Gossip Communication is a sufficient condition for the Tendermint algorithm to terminate, it is not strictly necessary, because the conditions to progress and terminate are evaluated over the messages of subsets of rounds executed, not all of them, and as new rounds are executed, messages in old rounds may be become obsolete and be ignored and forgotten, saving memory.
+In other words, the algorithm does not require all messages to be delivered, only messages that advance the state of the processes.
+
+## Node's state as a Tuple Space
+
+One way of looking at the information used by CometBFT nodes is as a distributed tuple space to which all nodes contribute with signed entries;
+to share a proposal, the proposer adds it to the tuple space and to vote for a proposal, the node adds the vote.
+Each update includes the author name and therefore they are all non-conflicting with each other and update forgery is prevented with signing each update.
+
+Since we are talking about an asynchronous distributed system, individual nodes can only maintain approximations of the tuple space.
+Nodes may broadcast the update operations to all nodes, including themselves, and, if the communication is reliable, as guaranteed by Gossip Communication, the tuple space will eventually converge.
+
+Otherwise, nodes may periodically compare their approximations with each other to identify and correct differences by adding missing entries, using some gossip/anti-entropy protocol.
+
+Updates are commutative from the point of view of the tuple space, even if they are not commutative from the Tendermint algorithm's point of view; however, nodes observing different approximations of the tuple space may decide at different point in time but cannot violate any correctness guarantees.
+And the eventual convergence of tuple space implies the eventual termination of the algorithm.
```

Comment:
Commutativity is not by itself sufficient to establish a CRDT, updates must also be associative and idempotent. And if you say that entries are removed from the set based on application-level properties of added (received) updates, then the set is not a CRDT by definition.

That's because elements which modify the value space based on prior state (which such values necessarily do) are not idempotent. Receiving elements {A B C} can produce a different outcome compared to receiving elements {A C B}.

Reply:
Thanks for you comment, but it will only make sense to try disprove what I intend here as a CRDT once I have formally defined the merge operator. Please keep in mind that this is a draft PR and is incomplete.